### PR TITLE
[WBP-198] Prevent conversation creation for non-fully connected graphs

### PR DESCRIPTION
**Pull request by battermann**
https://wearezeta.atlassian.net/browse/WPB-198

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: integration/test/Test/Conversation.hs [hunk 1]
```diff
@@ -135,3 +136,43 @@ testFederationStatus = do
...
+          >=> removeField "optSettings.setFederationDomainConfigs"
+          >=> setField "optSettings.setFederationDomainConfigsUpdateFreq" (Aeson.Number 1)
+  startDynamicBackends
+    [ def {dbBrig = setFederationConfig},
+      def {dbBrig = setFederationConfig},
+      def {dbBrig = setFederationConfig}
+    ]
+    $ \dynDomains -> do
+      domains@[domainA, domainB, domainC] <- pure dynDomains
+      sequence_ [Internal.createFedConn x (Internal.FedConn y "full_search") | x <- domains, y <- domains]
```
Comment (fisx) [inline integration/test/Test/Conversation.hs:153]:
```suggestion
      sequence_ [Internal.createFedConn x (Internal.FedConn y "full_search") | x <- domains, y <- domains, x /= y]
```

(it works without this change, but it's weird to register a domain to federate with itself.)

File: libs/wire-api/src/Wire/API/Conversation.hs [hunk 2]
```diff
@@ -313,6 +315,28 @@ instance ToSchema CreateGroupConversation where
...
+  }
+  deriving stock (Eq, Show, Generic)
+  deriving (Arbitrary) via (GenericUniform CreateConversationRejected)
+  deriving (ToJSON, FromJSON, S.ToSchema) via Schema CreateConversationRejected
+
+instance ToSchema CreateConversationRejected where
+  schema =
+    objectWithDocModifier
+      "CreateConversationRejected"
+      (description ?~ "A rejected conversation creation object extended with a list of non-federating backends")
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Conversation.hs:329]:
```suggestion
      (description ?~ "A rejected conversation creation object including a pair of remote backends that do not federate with each other")
```

File: services/galley/test/integration/API/Util.hs [hunk 3]
```diff
@@ -731,12 +732,14 @@ postConvWithRemoteUsers ::
   Maybe ClientId ->
   NewConv ->
   TestM (Response (Maybe LByteString))
-postConvWithRemoteUsers u c n =
+postConvWithRemoteUsers u c n = do
+  let mock =
+        ("get-not-fully-connected-backends" ~> NonConnectedBackends mempty)
+          <|> mockReply ()
```
Comment (fisx) [inline services/galley/test/integration/API/Util.hs:738]: i would make `postConvWithRemoteUsersGeneric` take a `mock`, and implement `postConvWithRemoteUsersGeneric`, `postConfWithRemoteUsersNonFullyConnected` in terms of that.

File: services/galley/test/integration/API.hs [hunk 4]
```diff
@@ -417,13 +419,15 @@ postConvWithRemoteUsersOk rbs = do
 
     liftIO $
       -- as many federated requests as remote backends, i.e., one per remote backend
-      Set.size (Set.fromList $ frTargetDomain <$> federatedRequests) @?= Set.size rbs
+      Set.size (Set.fromList $ frTargetDomain <$> federatedRequests') @?= Set.size rbs
+    for_ federatedRequests' print
```
Comment (fisx) [inline services/galley/test/integration/API.hs:423]: can you remove this noise and either not print anything, or use the test framework's output channels?

File: services/galley/test/integration/API.hs [hunk 5]
```diff
@@ -3330,7 +3343,7 @@ deleteRemoteMemberConvLocalQualifiedOk = do
               *> mockReply [mkProfile qEve (Name "Eve")]
           ]
   (convId, _) <-
-    withTempMockFederator' (mockedResponse <|> mockReply ()) $
+    withTempMockFederator' ("get-not-fully-connected-backends" ~> NonConnectedBackends mempty <|> mockedResponse <|> mockReply ()) $
```
Comment (fisx) [inline services/galley/test/integration/API.hs:3346]: general thought: should we create a few larger mock rule sets and give them names, so we can write many tests with them.  passing individual rules to each test feels like work.  not sure, though, maybe let's just see how this evolves.

File: services/galley/test/integration/API.hs [hunk 6]
```diff
@@ -2411,31 +2415,36 @@ postConvQualifiedFailBlocked = do
...
+  void $ withTempMockFederator' mock $ do
+    postConvQualified
+      (qUnqualified alice)
+      Nothing
+      defNewProteusConv
+        { newConvQualifiedUsers = [bob],
+          newConvTeam = Just (ConvTeamInfo tid)
+        }
+      !!! const 403 === statusCode
+  void $ withTempMockFederator' mock $ do
```
Comment (fisx) [inline services/galley/test/integration/API.hs:2439]:
```suggestion
```

File: services/galley/test/integration/API/Federation.hs [hunk 7]
```diff
@@ -934,8 +935,8 @@ sendMessage = do
           <!! const 201 === statusCode
 
   liftIO $ do
-    [galleyReq] <- case requests1 of
-      xs@[_] -> pure xs
+    galleyReq <- case requests1 of
+      [_, r] -> pure r
```
Comment (fisx) [inline services/galley/test/integration/API/Federation.hs:939]:
```suggestion
      [_, r] -> pure r -- (the first request is to `"get-not-fully-connected-backends"`.)
```

**Review by fisx — APPROVED** [review]
all good!  minus the inevitable nit-picks.  :)

File: integration/test/Test/Conversation.hs [hunk 8]
```diff
@@ -135,3 +136,43 @@ testFederationStatus = do
...
+        setField "optSettings.setFederationStrategy" "allowDynamic"
+          >=> removeField "optSettings.setFederationDomainConfigs"
+          >=> setField "optSettings.setFederationDomainConfigsUpdateFreq" (Aeson.Number 1)
+  startDynamicBackends
+    [ def {dbBrig = setFederationConfig},
+      def {dbBrig = setFederationConfig},
+      def {dbBrig = setFederationConfig}
+    ]
+    $ \dynDomains -> do
+      domains@[domainA, domainB, domainC] <- pure dynDomains
```
[the remark this answers is not in the document]
Comment (fisx) [reply integration/test/Test/Conversation.hs:152]: sorry, just saw that you're doing this on purpose.  i'm still opposed and would just turn off `-Wnon-exhaustive-patterns` (or whatever it's called), but it's your code!

[the remark this answers is not in the document]
Comment (battermann) [reply integration/test/Test/Conversation.hs:152]: It's kind of a convention we agreed on in the integration test suite. It is not ideal, but it makes it a bit easier on the call side.

File: services/galley/test/integration/API.hs (hunk 4 above)
Comment (battermann) [reply services/galley/test/integration/API.hs:423]: Oh, that was unintentional.

**Pull request merged**